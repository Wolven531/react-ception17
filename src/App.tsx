import React, { FC, useLayoutEffect } from "react";
import { Link, Route, useHistory } from "react-router-dom";
import type { History } from "history";
import "./App.css";

const loadPlugins = (history: History) => {
  console.log("adding init event listener");

  window.document.addEventListener(
    "init",
    ((e: CustomEvent<any>) => {
      const userLoadedEvt = new CustomEvent("user-loaded", {
        detail: {
          history,
        },
      });

      console.log("init event received, firing user-loaded");
      window.document.dispatchEvent(userLoadedEvt);
    }) as EventListener,
    {
      capture: true,
      once: false, // allow for multiple init events
    }
  );

  // !! in a real world app, below could be a dynamic load (e.g. from a URL)
  // !! rather than a local require invocation
  console.log("loading FakePluginLayout + FakeNav w/ require()");

  require("./FakePluginLayout");
  // require("./FakeNav")
};

export const App: FC = () => {
  const history = useHistory();

  useLayoutEffect(() => {
    console.log(`[App] loaded; about to loadPlugins`);

    loadPlugins(history);
  }, [history]);

  const handleClick = () => {
    console.log("[App] button was clicked");
  };

  return (
    <>
      <h1>App Host - React 17</h1>
      <Route path={"/"} exact>
        <div className="app" id="app">
          <h2>App Host root</h2>
          <Link to="/asdf">asdf (from App Host)</Link>
          <br />
          <button onClick={handleClick}>App Host button</button>
        </div>
      </Route>
      <Route path={"/asdf"}>
        <div>
          <h2>asdf from App Host</h2>
          <Link to="/">Home (from App Host)</Link>
        </div>
      </Route>
      <hr />
      <div id="placeholder-2"></div>
      <div id="placeholder-3"></div>
    </>
  );
};
