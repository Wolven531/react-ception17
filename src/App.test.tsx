import React from 'react'
import { render } from '@testing-library/react'
import { App } from './App'
import { MemoryRouter } from 'react-router'

describe('App', () => {
	it('renders view that matches snapshot', () => {
		const view = render(
			<MemoryRouter>
				<App />
			</MemoryRouter>,
		)

		expect(view.asFragment()).toMatchSnapshot()
	})
})
