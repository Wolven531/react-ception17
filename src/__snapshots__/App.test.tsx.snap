// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders view that matches snapshot 1`] = `
<DocumentFragment>
  <h1>
    FakePortal - React 17
  </h1>
  <div
    class="app"
    id="app"
  >
    <p>
      FakePortal root
    </p>
    <a
      href="/asdf"
    >
      asdf (from FakePortal)
    </a>
    <br />
    <button>
      FakePortal button
    </button>
  </div>
  <hr />
  <div
    id="placeholder-2"
  >
    <div
      class="FakePluginLayout"
    >
      <p>
        FakePluginLayout (injected)
      </p>
      <a
        href="/asdf"
      >
        asdf (from FakePluginLayout)
      </a>
      <br />
      <button>
        FakePluginLayout button
      </button>
    </div>
  </div>
  <div
    id="placeholder-3"
  />
</DocumentFragment>
`;
